tokio-util/src/sync/reusable_box.rs

```diff
@@ -122,32 +87,83 @@ impl<T> Future for ReusableBoxFuture<'_, T> {
     }
 }
 
-// The future stored inside ReusableBoxFuture<'_, T> must be Send since the
-// `new` and `set` and `try_set` methods only allow setting the future to Send
-// futures.
-//
-// Note that T is the return type of the future, so its not relevant for
-// whether the future itself is Send.
-unsafe impl<T> Send for ReusableBoxFuture<'_, T> {}
-
 // The only method called on self.boxed is poll, which takes &mut self, so this
 // struct being Sync does not permit any invalid access to the Future, even if
 // the future is not Sync.
 unsafe impl<T> Sync for ReusableBoxFuture<'_, T> {}
 
-// Just like a Pin<Box<dyn Future>> is always Unpin, so is this type.
-impl<T> Unpin for ReusableBoxFuture<'_, T> {}
+impl<T> fmt::Debug for ReusableBoxFuture<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ReusableBoxFuture").finish()
+    }
+}
+
+fn reuse_pin_box<T: ?Sized, U, O, F>(boxed: Pin<Box<T>>, new_value: U, callback: F) -> Result<O, U>
+where
+    F: FnOnce(Box<U>) -> O,
+{
+    let layout = Layout::for_value::<T>(&*boxed);
+    if layout != Layout::new::<U>() {
+        return Err(new_value);
+    }
+
+    // SAFETY: We don't ever construct a non-pinned reference to the old `T` from now on, and we
+    // always drop the `T`.
+    let raw: *mut T = Box::into_raw(unsafe { Pin::into_inner_unchecked(boxed) });
 
-impl<T> Drop for ReusableBoxFuture<'_, T> {
+    // When dropping the old value panics, we still want to call `callback` — so move the rest of
+    // the code into a guard type.
+    let guard = CallOnDrop::new(|| {
+        let raw: *mut U = raw.cast::<U>();
+        unsafe { raw.write(new_value) };
+
+        // SAFETY:
+        // - `T` and `U` have the same layout.
+        // - `raw` comes from a `Box` that uses the same allocator as this one.
+        // - `raw` points to a valid instance of `U` (we just wrote it in).
+        let boxed = unsafe { Box::from_raw(raw) };
+
+        callback(boxed)
+    });
+
+    // Drop the old value.
+    unsafe { ptr::drop_in_place(raw) };
+
+    // Run the rest of the code.
+    Ok(guard.into_inner()())
```

Comment:
IMO, using `catch_unwind` was a bit more straightforward than the guard type, as it involved less non-linear control flow...are there specific reasons why the `CallOnDrop` approach is superior?

Reply:
As far as I know, `catch_unwind` optimizes more poorly.

Reply:
I don't have an authoritative source for this, but the [`replace_with`](https://docs.rs/replace_with) docs say:

> though uses `Drop` instead of [`std::panic::catch_unwind()`](https://doc.rust-lang.org/nightly/std/panic/fn.catch_unwind.html) to react to unwinding, which avoids the optimisation barrier of calling the `extern "C" __rust_maybe_catch_panic()`. As such it's up to ∞x faster.

Another option though is to write it like this:

```rs
let try_run = || {
    unsafe { ptr::drop_in_place(raw) };
};
let finally = || {
    let raw: *mut U = raw.cast::<U>();
    unsafe { raw.write(new_value) };

    // SAFETY:
    // - `T` and `U` have the same layout.
    // - `raw` comes from a `Box` that uses the same allocator as this one.
    // - `raw` points to a valid instance of `U` (we just wrote it in).
    let boxed = unsafe { Box::from_raw(raw) };

    callback(boxed)
};
let guard = CallOnDrop::new(finally);
try_run();
Ok(guard.into_inner()())
```

Which somewhat avoids the non-local control flow issue. What do you think?

Reply:
I'm not really sure if the second closure makes it all that much clearer. I'm fine with the original implementation.

Thanks for the docs reference, though, that's really useful!